script/generate-version.js

```diff
@@ -1,10 +1,31 @@
 #!/usr/bin/env node
 
-const { writeFileSync } = require('fs');
+const { readFileSync, writeFileSync } = require('fs');
 const { gitDescribeSync } = require('git-describe');
 const path = require('path');
 
-const description = JSON.stringify(gitDescribeSync(), null, 2);
+let description;
+try {
+  description = JSON.stringify(gitDescribeSync(), null, 2);
+} catch (e) {
+  console.log(`Got error ${e} getting description of current version.`);
+  console.log('For an accurate version number, or for distribution, please use a full Git checkout of this repository.');
+
+  // Probably not a checkout. This is what you'll get if you use
+  // "Download ZIP" in the GitHub UI.
+  // Synthesize a version file
+  // from the current package.json.
+  const package = readFileSync(path.join(__dirname, '../package.json'));
+  const version = JSON.parse(package).version;
+  console.log(`Using synthetic build description with version ${version}.`)
+
+  description = JSON.stringify({
+    hash: 'gdeadbeef',
+    raw: 'v' + version,
+    semverString: version,
```

Comment:
I see `semverString` as just the version here where as in normal git repository `gitDescribeSync` will generate the `semverString` as something like `1.20.2+1.g42b39c4` (seems like the hash is getting attached to it). 

Also this is not adding all the keys like `semver`, `suffix` etc observed in the current generated `version.ts` will that affect the build as well as the publish process later? 

Also this is scoped down to error case so impact will be low if any I assume.

Reply:
By definition it's not possible for us to attach the commit hash of the current commit when we're not in a Git repo. The `gdeadbeef` is there as a sentinel value to support the hash accessor in `Versioning`.

The omitted keys are not used. All consumers of this data should go through `Versioning`.

The sole purpose for this commit is to allow someone to run `npm run build` on a downloaded copy of the repo and have it succeed rather than fail because there's no `.git` directory.

We expect real consumers to use a package manager (most) or to use a local clone (most of the rest), both of which will have correct, detailed version info.

Reply:
Cool got it thanks!